src/v/cluster/tests/partition_balancer_planner_test.cc

```diff
@@ -0,0 +1,243 @@
+
+    auto reassignments = planner.get_ntp_reassignments(hr, fm);
+    BOOST_REQUIRE_EQUAL(reassignments.size(), 0);
+}
+
+FIXTURE_TEST(
+  test_node_down_and_node_is_full_but_there_is_empty,
+  partition_balancer_fixture) {
+    vlog(logger.debug, "test_node_down_and_node_is_full_but_there_is_empty");
+    allocator_register_nodes(5);
+    create_topic("topic-1", 1, 3);
+    auto hr = create_health_report();
+    auto fm = create_follower_metrics();
+
+    // Set one node unavailable
+    fm[0].last_heartbeat = raft::clock_type::now() - std::chrono::seconds(5min);
+
+    // Set one node full
+    hr.node_reports[3].local_state.disks[0].free = nearly_full_node_free_size;
+
+    vlog(logger.debug, "start planner");
+    auto reassignments = planner.get_ntp_reassignments(hr, fm);
+    BOOST_REQUIRE_EQUAL(reassignments.size(), 1);
+
+    std::unordered_set<model::node_id> expected_nodes(
+      {model::node_id(1), model::node_id(2), model::node_id(4)});
+
+    auto new_replicas = reassignments.front()
+                          .allocation_units.get_assignments()
+                          .front()
+                          .replicas;
+    BOOST_REQUIRE_EQUAL(new_replicas.size(), expected_nodes.size());
+    for (const auto r : new_replicas) {
+        BOOST_REQUIRE(expected_nodes.contains(r.node_id));
+    }
+}
+
+FIXTURE_TEST(test_node_down_and_node_is_full, partition_balancer_fixture) {
+    vlog(logger.debug, "test_node_down_and_node_is_full");
+    allocator_register_nodes(4);
+    create_topic("topic-1", 1, 3);
+    auto hr = create_health_report();
+    auto fm = create_follower_metrics();
+
+    // Set one node unavailable
+    fm[0].last_heartbeat = raft::clock_type::now() - std::chrono::seconds(5min);
+
+    // Set one node full
+    hr.node_reports[3].local_state.disks[0].free = nearly_full_node_free_size;
+
+    vlog(logger.debug, "start planner");
+    auto reassignments = planner.get_ntp_reassignments(hr, fm);
+    BOOST_REQUIRE_EQUAL(reassignments.size(), 0);
+}
+
+FIXTURE_TEST(test_move_from_full_node, partition_balancer_fixture) {
+    vlog(logger.debug, "test_move_from_full_node");
+    allocator_register_nodes(4);
+    create_topic("topic-1", 1, 3);
+    auto hr = create_health_report();
+    auto fm = create_follower_metrics();
+
+    // Set one node full
+    hr.node_reports[0].local_state.disks[0].free = full_node_free_size;
+
+    auto reassignments = planner.get_ntp_reassignments(hr, fm);
+    BOOST_REQUIRE_EQUAL(reassignments.size(), 1);
+
+    std::unordered_set<model::node_id> expected_nodes(
+      {model::node_id(1), model::node_id(2), model::node_id(3)});
+
+    auto new_replicas = reassignments.front()
+                          .allocation_units.get_assignments()
+                          .front()
+                          .replicas;
+    BOOST_REQUIRE_EQUAL(new_replicas.size(), expected_nodes.size());
+    for (const auto r : new_replicas) {
+        BOOST_REQUIRE(expected_nodes.contains(r.node_id));
+    }
```

Comment:
it would be nice to have a helper to check reassignments against the expected value